examples/tutorials/core_api/5_profiling.py

```diff
@@ -0,0 +1,40 @@
+"""
+Stage 1: Let's create the core_context, and use it to start logging metrics.
+This is just a few lines of code, and we'll be able to see the results in the
+Determined WebUI.
+"""
+
+import logging
+import sys
+import time
+
+# NEW: import determined
+import determined as det
+from determined import profiler
+
+def main(core_context, increment_by):
+    core_context.profiler.profiler.set_training(True)
+    for batch in range(100):
+        with core_context.profiler.profiler.record_timing("train_batch.backward"):
+            steps_completed = batch + 1
+            time.sleep(1)
+            logging.info(f"batch is now {batch}")
+        core_context.profiler.profiler.update_batch_idx(batch)
```

Comment:
it would be very cool to offer some sort of pass-thru iterator that counted batches instead of requiring this sorta weird call.

```python
for batch in profiler.count_batches(range(100), start=0):
    ...
```

Reply:
i thought about this too. we need to know when to step, so either we control the iterator, or users tell us when to step. i agree calling update_batch_idx is a bit weird, but providing some sort of iterator also gets weird when you have a dataloader. what if i want to iterate through batches using enumerate(dataloader)? i feel like it makes more sense to just call step yourself. what if we called it `profiler.step_batch()` instead? no args, profiler keeps track of current batch step.